flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/LocalInputChannel.java

```diff
@@ -168,6 +175,12 @@ public void run() {
 	Optional<BufferAndAvailability> getNextBuffer() throws IOException, InterruptedException {
 		checkError();
 
+		BufferAndAvailability bufferAndAvailability = getNextRecoveredStateBuffer();
```

Comment:
I am neutral for this option, because the similar way really existed in many other places before. E.g. we have `BufferStorage` in `CheckpointedInputGate` for caching the blocked channels' buffers, then while `getNextBuffer` we also need to check whether there are  any pending buffers to be read from `BufferStorage` firstly.

I absolutely agree that it would be better to not have different paths, but I also think it is not so bad if no other easy options.